File: ruby/clock/clock.rb
# frozen_string_literal: true

class Clock
  def initialize(hour: 0, minute: 0)
    @total_minutes = (hour * 60 + minute) % (24 * 60)
  end

  def hour
    total_minutes / 60
  end

  def minute
    total_minutes % 60
  end

  def to_s
    format '%02i:%02i', hour, minute
  end

  def inspect
    "#<#{self.class.name}: #{self}>"
  end

  attr_reader :total_minutes

  def ==(other)
    total_minutes == other.total_minutes
  end

  def +(other)
    Clock.new(minute: total_minutes + other.total_minutes)
  end

  def -(other)
    Clock.new(minute: total_minutes - other.total_minutes)
  end
end

File: python/clock/clock.py
# -*- coding: utf-8 -*-

class Clock:
    def __init__(self, hour: int, minute: int, second: int = 0):
        self.hour = hour
        self.minute = minute
        self.second = second

        if self.minute >= 60 or self.minute < 0:
            hours_to_add, self.minute = divmod(self.minute, 60)
            self.hour += hours_to_add

        if self.hour >= 24 or self.hour < 0:
            self.hour %= 24

    def __str__(self):
        return '%02i:%02i' % (self.hour, self.minute)

    def __eq__(self, other: 'Clock'):
        return self.hour == other.hour and self.minute == other.minute

    def add(self, other: int):
        "Return a Clock with `other` more minutes than self."
        return Clock(self.hour, self.minute + other)
